plugins/Synchro/SynchroSynth.cpp

```diff
@@ -0,0 +1,565 @@
+//harmonic: how strong the magic harmonics are
+static inline float SynchroWaveform(float x, float drive, float sync, float chop, float harmonic)
+{
+	return fastTanh(trih(F_PI_2 + x * sync, harmonic) * drive / powf(F_2PI / (F_2PI - x), chop));
+}
+
+//SynchroNote empty constructor
+SynchroNote::SynchroNote(NotePlayHandle * nph) :
+	nph(nph)
+{
+
+}
+
+void SynchroNote::nextStringSample(sampleFrame &outputSample, sample_rate_t sample_rate,
+	const float & modulationStrength, const float & modulationAmount, const float & harmonics,
+	const SynchroOscillatorSettings & carrier, const SynchroOscillatorSettings & modulator)
+{
+	float freqToSampleStep = F_2PI / sample_rate;
+	//How long (in frames) this note has been playing
+	float NoteTime = nph->totalFramesPlayed();
+
+	//Modulator is calculated first because it is used by the carrier
+	//Find position in modulator waveform
+	modulatorSampleIndex += freqToSampleStep * DetuneOctaves(nph->frequency(), modulator.Detune);
+	//Make sure phase is always between 0 and 2PI
+	while (modulatorSampleIndex >= F_2PI) { modulatorSampleIndex -= F_2PI; }
+
+	//Get modulator waveform at position
+	float modulatorSample = SynchroWaveform(modulatorSampleIndex,
+		modulator.Drive, modulator.Sync, modulator.Chop, harmonics) * (modulationStrength * modulationAmount);
+
+	//Modulator envelope
+	//Using sample_rate here gives an incorrect envelope length for some reason
+	float modulatorEnvelope;
+	float modulatorAtk = modulator.Attack * 0.001 * Engine::mixer()->processingSampleRate();
+	float modulatorDec = modulator.Decay * 0.001 * Engine::mixer()->processingSampleRate();
+	float modulatorRel = modulator.Release * 0.001 * Engine::mixer()->processingSampleRate();
+	if (NoteTime < modulatorAtk) //Attack
+	{
+		//Linear attack sounds better
+		modulatorEnvelope = lerp(0, 1, NoteTime / modulatorAtk);
+	}
+	else if (NoteTime <= modulatorAtk + modulatorDec) //Decay
+	{
+		//But decay needs to be exponential
+		modulatorEnvelope = expInterpol(1, modulator.Sustain, NoteTime / (modulatorAtk + modulatorDec));
+	}
+	else //Sustain
+	{
+		modulatorEnvelope = modulator.Sustain;
+	}
+	if (nph->isReleased()) //Release is done separately so the volume doesn't jump when released prematurely
+	{
+		float releaseProgress = (nph->releaseFramesDone() < modulatorRel) ? nph->releaseFramesDone() : modulatorRel;
+		modulatorEnvelope *= expInterpol(1, 0, releaseProgress / modulatorRel);
+	}
+
+	//The final sample Synchro uses for phase modulation
+	float pmSample = modulatorSample * modulatorEnvelope * SYNCHRO_PM_CONST;
+
+	//Find position in carrier waveform
+	carrierSampleIndex += freqToSampleStep * DetuneOctaves(nph->frequency(), carrier.Detune);
+	//Make sure phase is always between 0 and 2PI
+	while (carrierSampleIndex >= F_2PI) { carrierSampleIndex -= F_2PI; }
+	//Extra check since phase modulation can push it below 0
+	while (carrierSampleIndex < 0) { carrierSampleIndex += F_2PI; }
+	//Carrier envelope
```

Comment:
It looks like the code fragment for the carrier envelope is similar to modulator's one, isn't it possible to reduce duplication?

Reply:
Good call! I made a little inline function called `wrapPhase`. If there's an even better way to to this, please let me know!

Reply:
I think I meant the whole logic for calculating `modulatorEnvelope` and `carrierEnvelope`.

Reply:
That logic has now been moved to an `ADSR()` function.